internal/p2p/applicator.go

```diff
@@ -310,14 +315,22 @@ func (b *Applicator) handleApplyBlock(request *blockApplicationRequest) {
 		err = p2perrors.ErrBlockIrreversibility
 	} else {
 		_, err = b.rpc.ApplyBlock(request.ctx, request.block)
+		if err == nil {
+			b.cache.CheckBlock(request.block)
+		}
 	}
 
 	request.errChan <- err
 	close(request.errChan)
 }
 
 func (b *Applicator) handleApplyTransaction(request *transactionApplicatorRequest) {
-	_, err := b.rpc.ApplyTransaction(request.ctx, request.trx)
+	var err error
+	if b.cache.CheckTransactions(request.trx) == 0 {
+		_, err = b.rpc.ApplyTransaction(request.ctx, request.trx)
+	} else {
+		err = p2perrors.ErrDuplicateTransaction
```

Comment:
Do we want to return an error here? This will mean that if we hear about a transaction via gossip that was already seen in a block that we will apply error score to that peer. Previously, we just accepted the transaction via gossip without applying it. I think that is the correct behavior.